File: 프로그래머스/unrated/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
class Solution {
    public int solution(int[] topping) {
        int answer = 0;
        Set<Integer> bro = new HashSet<>();
        Map<Integer, Integer> littleBro = new HashMap<>();
        for(int top : topping) {
            littleBro.put(top, littleBro.getOrDefault(top, 0) + 1);
        }

        for(int top : topping) {
            littleBro.computeIfPresent(top, (key, value) -> value-1);
            bro.add(top);

            if(littleBro.get(top) == 0) {
                littleBro.remove(top);
            }

            if(bro.size() == littleBro.size()) {
                answer++;
            }
        }

        return answer;
    }
}

File: 프로그래머스/unrated/132265. 롤케이크 자르기/롤케이크 자르기.py
from collections import Counter

def solution(topping):
    answer = 0
    bro = Counter(topping)
    little = set()
    
    for t in topping:
        bro[t] -= 1
        little.add(t)
        
        if bro[t] == 0:
            del bro[t]
            
        if len(bro) == len(little):
            answer+=1
            
    return answer

# 시간초과
def solution_fail(topping):
    answer = 0
    for i in range(1,len(topping)-1):
        little = len(set(topping[:i]))
        bro = len(set(topping[i:]))
        
        if little == bro:
            answer += 1
            
    return answer